### PR TITLE
Support unconstrained randomization for associative array and queue (#4625)

### PR DESCRIPTION
**Pull request by YilouWang**
This patch adds support for unconstrained randomization for associative arrays and queues. Thanks to @udayaa-init  for his kind help.

Following the previous PR #5415 , this patch became much simpler. With this change, unconstrained randomization is now supported for all array types basically. 

Moving forward, we will address this issue #5455  and test more complex nested array structures.

### CODE REVIEW

File: src/V3Randomize.cpp [hunk 1]
```diff
@@ -1369,17 +1376,24 @@ class RandomizeVisitor final : public VNVisitor {
             return new AstForeach{fl, randLoopVarp, newRandStmtsp(fl, tempElementp, nullptr)};
         };
         AstNodeExpr* tempElementp = nullptr;
-        while (VN_CAST(tempDTypep, DynArrayDType) || VN_CAST(tempDTypep, UnpackArrayDType)) {
+        while (VN_CAST(tempDTypep, DynArrayDType) || VN_CAST(tempDTypep, UnpackArrayDType)
+               || VN_CAST(tempDTypep, AssocArrayDType) || VN_CAST(tempDTypep, QueueDType)) {
```
Comment (wsnyder) [inline src/V3Randomize.cpp:1380]:
```suggestion
        while (VN_IS(tempDTypep, DynArrayDType) || VN_IS(tempDTypep, UnpackArrayDType)
               || VN_IS(tempDTypep, AssocArrayDType) || VN_IS(tempDTypep, QueueDType)) {
```

File: src/V3Randomize.cpp [hunk 2]
```diff
@@ -1369,17 +1376,24 @@ class RandomizeVisitor final : public VNVisitor {
...
+                  : VN_CAST(tempDTypep, UnpackArrayDType) ? static_cast<AstNodeExpr*>(
+                        new AstArraySel{fl, tempElementp ? tempElementp : exprp,
+                                        new AstVarRef{fl, newRandLoopIndxp, VAccess::READ}})
+                  : VN_CAST(tempDTypep, AssocArrayDType) ? static_cast<AstNodeExpr*>(
+                        new AstAssocSel{fl, tempElementp ? tempElementp : exprp,
+                                        new AstVarRef{fl, newRandLoopIndxp, VAccess::READ}})
+                  : VN_CAST(tempDTypep, QueueDType) ? static_cast<AstNodeExpr*>(new AstCMethodHard{
+                        fl, tempElementp ? tempElementp : exprp, "atWriteAppend",
+                        new AstVarRef{fl, newRandLoopIndxp, VAccess::READ}})
+                                                    : nullptr;
```
Comment (wsnyder) [inline src/V3Randomize.cpp:1396]: I think this would be clearer as a series of `if` and assignments.  Then the casts shouldn't be needed.

**Review by wsnyder — COMMENTED** [review]
Please make sure you have recent master merged in.

Can you please comment-in the rand's listed as issue-4625 in t_uvm_pkg_todo.vh?

**Comment by wsnyder** [comment]
BTW I think this should be extended cover dyn arrays too, as they are nearly identical to queues.

**Comment by YilouWang** [comment]
Thanks for the comments.

Dynamic array has been already supported with PR #5415 . Only assoc and queue left, and this patch supports both of them.

**Comment by wsnyder** [comment]
Ok, great, so then should be ok to comment-in the uvm_todo rand's.

**Comment by YilouWang** [comment]
Hello, @wsnyder 
I just checked the code in uvm_todo rand's which is with constraints. 
This patch supports **unconstrained randomization** for associative arrays and queues. Therefore, it doesn’t currently help with the uvm_todo part of the code. I will suport constrained randomization part in future PR.

Based on the uvm_todo part's test result, rand field seems supported. Error shows "`fifo.size()` not supported, treating as state..." warning message. Maybe I could investigate it and fix it later in other pull request.

**Comment by wsnyder** [comment]
Ok, thanks, if you can look at  a future pull for that UVM rand, that would be appreciated.

**Pull request merged**